### PR TITLE
CIP-0035 | Redraft to handle Plutus Core language versions

### PR DESCRIPTION
**Pull request by michaelpj**
- Include discussion of Plutus Core language versions and how they change.
- Use "ledger language" instead of "language version" or "ledger language version" to avoid confusion
- Remove the reference to the 'Draft' status.
- Minor prose improvements.

The actual Plutus Core language version was not discussed previously, as it hadn't really seen much use. However we might want to change Plutus Core itself in the medium-term future, so we should make it clear how to do that.

I also tried to clear up some confusing language - the use of 'language version' for the ledger languages was just too confusing, especially in the presence of _another_ language version! So I've opted to just refer to 'ledger languages', which I think is clearer.

I'm planning to do another PR to do more adjustments to make this fit the new CIP-001 better, but I wanted to do this first as it has some substantive changes.

### CODE REVIEW

**Review by rphair — COMMENTED** [review]
As far as I can tell this also completes https://github.com/cardano-foundation/CIPs/issues/420.  It all looks good to me but @michaelpj could you tag a couple more reviewers who would be willing to go over the text before we merge?

**Comment by michaelpj** [comment]
I'll ask a couple of the Plutus team to check it over.

File: CIP-0035/README.md
Comment (kwxm) [inline CIP-0035/README.md:198]: In some cases an implementation using the existing features might be difficult or impossible. For example, Implementing a complex cryptographic function using the basic integer operations might be a very large amount of work for something that would clearly be inferior to a C implementation.

Maybe something like "If possible/feasible, a comparison ..."?

File: CIP-0035/README.md
Comment (kwxm) [inline CIP-0035/README.md:270]: ... also because there could be an un-triggered validator on the chain that relies on the old behaviour?  Maybe the code is law and that's the owner's problem, but they'd have a reasonable grievance if the behaviour changed from what was advertised when the code was written.

File: CIP-0035/README.md
Comment (kwxm) [inline CIP-0035/README.md:192]: There are certain extra conditions that we'll require to hold: for example the builtin must always terminate and must never throw any exceptions, and its behaviour should be predictable (there shouldn't be any worst-case behaviour that's significantly worse than the costing function would predict).  I don't think we have a formal list of such things yet, but we'll need one eventually and it's not clear to me whether it should go in this CIP or somewhere else.  Maybe we need a builtin implementer's guide.

File: CIP-0035/README.md
Comment (kwxm) [inline CIP-0035/README.md:267]: What do we do if we discover that some builtin contains a bug that can compromise the chain, like a division by zero error in a C library?

Also, what do we do if there's a change in the behaviour of the (Haskell or C) library function underlying a builtin at some point in the future? Would we have to maintain a version of the library function reproducing the old behaviour in perpetuity and use it for the execution of existing scripts depending on the old behaviour?

**Review by kwxm — APPROVED** [review]
I've added a few comments, maybe not all of which are directly related to this CIP.

Also, "implementer" or "implementor"?  I'm never sure about that myself, but Google and my dictionary seem to favour the former.

All in all I think this does a pretty good job of explaining the complexities of how the different versions of different things all fit together: that's quite confusing so it's very helpful to have it all clearly explained in one place.

**Comment by kwxm** [comment]
A couple of bonus comments.  Near the beginning it says

> Plutus Core has a number of builtin types, such as integers, and builtin functions, such as integer addition. Builtin functions provide access to functionality that would be difficult or expensive to implement in Plutus Core code.

That's kind of self-contradictory.  Maybe it could say something like "... expensive to implement using the basic constructs of Plutus Core"?  It might also be worth having a sentence saying that the non-builtin part of Plutus Core is something like the untyped lambda calculus.

Just below that it says

> Builtin functions can operate only over builtin types.

That's not strictly true (cf `ifThenElse`, where the branches are just terms), but I'm not sure how to explain the true situation concisely.

File: CIP-0035/README.md
Comment (michaelpj) [reply CIP-0035/README.md:68]: I blame markdown. It's hard to do reliable section links so I just didn't try. I guess I can try again.

File: CIP-0035/README.md
Comment (michaelpj) [reply CIP-0035/README.md:193]: I think I'm deliberately leaving that a little vague because I'm not sure. I think I want them to _talk_ about it but I think we might want more or less depending on the proposal :thinking:

File: CIP-0035/README.md
Comment (michaelpj) [reply CIP-0035/README.md:192]: I'll add some of that.

File: CIP-0035/README.md
Comment (michaelpj) [reply CIP-0035/README.md:267]: I think that's all out of scope of this CIP (since I would say that's "emergency measures"), but
1. No script can ever depend on behaviour that crashes the node. So we can deal with such things. If there's wrong but non-crashing behaviour... we need to preserve that.
2. Yes, if there's a change in the behaviour of the underlying function we'll need to maintain a version with the old behaviour. I guess that's another criteria for builtins: the behaviour should be as clearly specified as possible, so it won't change.

File: CIP-0035/README.md
Comment (michaelpj) [reply CIP-0035/README.md:193]: Also: I think those benchmarks would be part of the implementation, which is separate from the CIP.

**Review by thealmarty — APPROVED** [review]

**Review by SebastienGllmt — APPROVED** [review]

**Review by rphair — APPROVED** [review]
Looks like @michaelpj last commit addresses last round of feedback but will wait for @KtorZ to approve / merge.

**Review by KtorZ — APPROVED** [review]
Happy with the update, although this raises an interesting follow-up question:

Why is there a Plutus-core language version if it's always set to 1.0.0 and, according to this new update, is never going to change?

**Comment by michaelpj** [comment]
> Why is there a Plutus-core language version if it's always set to 1.0.0 and, according to this new update, is never going to change?

Hmmm, the whole point of this update was to convey that it _will_ change if we change any of the constructs of Plutus Core itself. Which bit made it sound like it won't change?

File: CIP-0035/README.md
Comment (zliu41) [inline CIP-0035/README.md:144]: Doesn't this depend on what changed in the new LV? Imagine a new LV that removes builtin type `Data` (by the way, do we still need `Data` once we have sums of products?). This would lead to an unavoidable change in the ledger-script interface, necessitating a new LL.

**Comment by KtorZ** [comment]
> @michaelpj Which bit made it sound like it won't change?

Every mention of the Plutus language being removed 😅? There's only mention of the ledger language version now (`LL`), which as I understand it refers to the language tag that the ledger uses to distinguish between different Plutus version. It is thus different from the on-chain program version that encapsulate every UPLC programs, and that has stayed to `1.0.0` so far.

**Comment by michaelpj** [comment]
> Every mention of the Plutus language being removed sweat_smile? There's only mention of the ledger language version now (LL), which as I understand it refers to the language tag that the ledger uses to distinguish between different Plutus version. It is thus different from the on-chain program version that encapsulate every UPLC programs

Now we have the LV for the "program version" and LL for the different Plutus ledger languages. And both can change - there's an explicit type of change for adding a new LV.

File: CIP-0035/README.md
Comment (michaelpj) [reply CIP-0035/README.md:144]: Sure, but then you're also including a change to the ledger-script interface. If you do multiple changes obviously you need to biggest kind of release.

Comment (zliu41) [reply CIP-0035/README.md:144]: Fair enough.

**Comment by kwxm** [comment]
There seems to be something wrong with some of the links in the [version](https://cips.cardano.org/cips/cip35/) that's currently publicly visible.

CIP35

**Comment by rphair** [comment]
thanks @kwxm - this is a known issue (footnotes not working) (https://github.com/cardano-foundation/CIPs/issues/435) which we'll be able to resolve when most CIPs are edited for the new format & we can overhaul the web site generator.

**Pull request merged**
